icu4c/source/common/cjmlbe.cpp

```diff
@@ -0,0 +1,513 @@
+    /**
+     * Default constructor.
+     *
+     * @param ch A Unicode character
+     * @param index The unicode block index of the character.
+     */
+    Element(UChar32 character, const UnicodeString &idx) : ch(character), index(idx) {}
+
+    UChar32 ch;
+    UnicodeString index;
+};
+
+CjMLBreakEngine::CjMLBreakEngine(const UnicodeSet &digitOrOpenPunctuationOrAlphabetSet,
+                                 const UnicodeSet &closePunctuationSet, UErrorCode& status)
+    : fDigitOrOpenPunctuationOrAlphabetSet(digitOrOpenPunctuationOrAlphabetSet),
+      fClosePunctuationSet(closePunctuationSet) {
+    if (U_FAILURE(status)) {
+        return;
+    }
+    loadMLmodel(status);
+}
+
+CjMLBreakEngine::~CjMLBreakEngine() {
+}
+
+namespace {
+    const char16_t INVALID = u'|';
+}
+
+static UnicodeString* concatChar(const char16_t *str, const UChar32 *arr, int32_t length,
+                                 UErrorCode &status) {
+    LocalPointer<UnicodeString> result(new UnicodeString(str), status);
+    if (U_FAILURE(status)) {
+        return nullptr;
+    }
+    int32_t idx = 0;
+    for (int i = 0; i < length; i++) {
+        result->append(arr[idx++]);
+    }
+    return result.orphan();
+}
+
+static UnicodeString* cloneString(const UnicodeString &str, UErrorCode &status) {
+    LocalPointer<UnicodeString> result(str.clone(), status);
+    return result.orphan();
+}
+
+static void deleteElement(void *obj) {
+    delete static_cast<Element *>(obj);
+}
+
+static bool isValid(const UnicodeString &s) {
+    return s.length() != 1 || s[0] != INVALID;
+}
+
+
+int32_t CjMLBreakEngine::divideUpDictionaryRange(UText *inText, int32_t rangeStart,
+                                                 int32_t rangeEnd, UVector32 &foundBreaks,
+                                                 const UnicodeString &inString,
+                                                 const LocalPointer<UVector32> &inputMap,
+                                                 UErrorCode &status) const {
+
+    if (U_FAILURE(status)) {
+        return 0;
+    }
+    if (rangeStart >= rangeEnd) {
+        status = U_ILLEGAL_ARGUMENT_ERROR;
+        return 0;
+    }
+
+    UVector32 boundary(inString.countChar32() + 1, status);
+    if (U_FAILURE(status)) {
+        return 0;
+    }
+    int32_t numBreaks = 0;
+    UChar32 ch;
+    UnicodeString index;
+    UVector elementList(&deleteElement, nullptr ,6, status);
+    if (U_FAILURE(status)) {
+        return 0;
+    }
+
+    for (int32_t i = 1; i < inString.length(); i++) {
+        if (elementList.size() > 0) {
+            // Remove the first element and append a new element which starts from index-4 (i=2)"+2"
+            elementList.removeElementAt(0);
+            ch = (i + 2) < inString.length() ? inString.char32At(i + 2) : INVALID;
+            index = (ch != INVALID) ? getUnicodeBlockIndex(ch, status) : UnicodeString(INVALID);
+            elementList.adoptElement(new Element(ch, index), status);
+        } else {
```

Comment:
I think the way you call initElementList is very strange in this function, you place inside a loop which mislead me to believe it will be call multiple times until I realize it is in the else block of elementList.size() > 0). How about change the whole loop like


```
UVector elementList(&deleteElement, nullptr ,6, status);
initElementList(inString, elementList, status);
if (U_FAILURE(status)) return 0;
for (int32_t i = 1; i < inString.length() && U_SUCCESS(status); i++) {
      U_ASSERT(elementList.size() > 0);
      evaluateBreakpoint (elementList, i, numBreaks, boundary, status);
      if (i + 1 >= inString.length()) break;
      // Remove the first element and append a new element which starts from index-4 (i=2)"+2"
      elementList.removeElementAt(0);
      ch = (i + 3) < inString.length() ? inString.char32At(i + 3) : INVALID;
      index = (ch != INVALID) ? getUnicodeBlockIndex(ch, status) : UnicodeString(INVALID);
      elementList.adoptElement(new Element(ch, index), status);
  }
}
if (U_FAILURE(status)) return 0;
```

Reply:
Done. Thanks for the suggestion.